src/data/dataloader.cpp

```diff
@@ -0,0 +1,693 @@
+
+        // Identifier
+        uint8_t idLen;
+        file.read(reinterpret_cast<char*>(&idLen), sizeof(uint8_t));
+        e.identifier.resize(idLen);
+        file.read(e.identifier.data(), idLen);
+
+        // Text
+        uint16_t len;
+        file.read(reinterpret_cast<char*>(&len), sizeof(uint16_t));
+        e.text.resize(len);
+        file.read(e.text.data(), len);
+
+        result.entries.push_back(e);
+    }
+
+    return result;
+}
+
+void saveCachedFile(const Labelset& labelset, std::filesystem::path path) {
+    std::ofstream file(path, std::ofstream::binary);
+
+    file.write(reinterpret_cast<const char*>(&LabelCacheFileVersion), sizeof(int8_t));
+
+    //
+    // Storage text color
+    checkSize<int16_t>(labelset.textColorIndex, "Too high text color");
+    int16_t textColorIdx = static_cast<int16_t>(labelset.textColorIndex);
+    file.write(reinterpret_cast<const char*>(&textColorIdx), sizeof(int16_t));
+
+    //
+    // Storage text lines
+    checkSize<uint32_t>(labelset.entries.size(), "Too many entries");
+    uint32_t nEntries = static_cast<uint32_t>(labelset.entries.size());
+    file.write(reinterpret_cast<const char*>(&nEntries), sizeof(uint32_t));
+    for (const Labelset::Entry& e : labelset.entries) {
+        file.write(reinterpret_cast<const char*>(&e.position.x), sizeof(float));
+        file.write(reinterpret_cast<const char*>(&e.position.y), sizeof(float));
+        file.write(reinterpret_cast<const char*>(&e.position.z), sizeof(float));
+
+        // Identifier
+        checkSize<uint8_t>(e.identifier.size(), "Identifier too long");
+        uint8_t idLen = static_cast<uint8_t>(e.identifier.size());
+        file.write(reinterpret_cast<const char*>(&idLen), sizeof(uint8_t));
+        file.write(e.identifier.data(), idLen);
+
+        // Text
+        checkSize<uint16_t>(e.text.size(), "Text too long");
+        uint16_t len = static_cast<uint16_t>(e.text.size());
+        file.write(reinterpret_cast<const char*>(&len), sizeof(uint16_t));
+        file.write(e.text.data(), len);
+    }
+}
+
+Labelset loadFileWithCache(std::filesystem::path filePath) {
+    return internalLoadFileWithCache<Labelset>(
+        filePath,
+        std::nullopt,
+        &loadFile,
+        &loadCachedFile,
+        &saveCachedFile
+    );
+}
+
+} // namespace label
+
+namespace color {
+
+ColorMap loadFile(std::filesystem::path path, std::optional<DataMapping>) {
+    ghoul_assert(std::filesystem::exists(path), "File must exist");
+
+    std::ifstream file(path);
+    if (!file.good()) {
+        throw ghoul::RuntimeError(fmt::format("Failed to open colormap file {}", path));
+    }
+
+    std::string extension = path.extension().string();
+    std::transform(extension.begin(), extension.end(), extension.begin(),
+        [](unsigned char c) { return std::tolower(c); });
+
+    ColorMap res;
+    if (extension == ".cmap") {
+        res = speck::loadCmapFile(path.string());
```

Comment:
The `loadCmapFile` should probably take a `std::filesystem::path` as an argument, in which case this could just be `path`

Reply:
Interestingly, it already does.  But good catch, I'll remove all the extra `.string()`s here